File: misc/BinarySearch/binarySearch.py
def binarySearch(nums, target):
    low, high = 0, len(nums) - 1
    while low <= high:
        mid = low + (high-low) // 2
        if target == nums[mid]:
            return mid
        elif target > nums[mid]:
            low = mid + 1
        else:
            high = mid - 1
    return -1

def binarySearch2(nums, target):
    low, high = 0, len(nums)
    while low < high:
        mid = low + (high - low) // 2
        if target == nums[mid]:
            return mid
        elif target > nums[mid]:
            low = mid + 1
        else:
            high = mid
    return -1

if __name__ == "__main__":
    s = [2,4,5,6,9,11, 13, 40, 40, 102]
    for i in range(len(s)):
        print(binarySearch(s, s[i]))
    assert binarySearch(s, 10) == -1

    for i in range(len(s)):
        print(binarySearch2(s, s[i]))
    assert binarySearch2(s, 10) == -1

File: leetcode/704-BinarySearch/binarySearch.cc
//    Given a sorted (in ascending order) integer array nums of n elements and a target value, write a function to search target in nums. If target exists, then return its index, otherwise return -1.
//
//
//    Example 1:
//
//    Input: nums = [-1,0,3,5,9,12], target = 9
//    Output: 4
//    Explanation: 9 exists in nums and its index is 4
//
//    Example 2:
//
//    Input: nums = [-1,0,3,5,9,12], target = 2
//    Output: -1
//    Explanation: 2 does not exist in nums so return -1
//
//
//    Note:
//
//    You may assume that all elements in nums are unique.
//    n will be in the range [1, 10000].
//    The value of each element in nums will be in the range [-9999, 9999].

#include <vector>
#include <cassert>

using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size()-1;
        while (left <= right)
        {
            // invariant: index \in [left, right]
            int mid = left + (right - left) / 2;
            if(nums[mid] > target)
            {
                right = mid - 1;
            }
            else if (nums[mid] < target)
            {
                left = mid + 1;
            }
            else
            {
                return mid;
            }
        }
        return -1;
    }
};

using ptr2search  = int (Solution::*)(vector<int>&, int);

void test(ptr2search pfcn)
{
    Solution sol;
    vector<int> nums = {-1,0,3,5,9,12};
    int target = 9;
    assert((sol.*pfcn)(nums,9) == 4);
    nums = {-1,0,3,5,9,12};
    target = 2;
    assert((sol.*pfcn)(nums,2) == -1);
}

int main()
{
    ptr2search pfcn = &Solution::search;
    test(pfcn);
}